build-system/tasks/compile.js

```diff
@@ -14,80 +14,101 @@
-    // This is needed because we currently use only very basic optimizations
-    // in closure compiler and it doesn't minify global variables at this
-    // stage.
-    return gulp.src(['build/cc-amp.js'])
-        .pipe(uglify({
-          preserveComments: 'some'
-        }))
-        .pipe(rename('cc-v0.js'))
-        .pipe(gulp.dest('dist'))
+    wrapper += '\n//# sourceMappingURL=' +
+        outputFilename + '.map\n';
+    /*eslint "google-camelcase/google-camelcase": 0*/
+    return gulp.src([
+      '3p/**/*.js',
+      'ads/**/*.js',
+      'extensions/**/*.js',
+      'build/**/*.js',
+      '!build/polyfills.js',
+      'src/**/*.js',
+      '!third_party/babel/custom-babel-helpers.js',
+      // Exclude since it's not part of the runtime/extension binaries.
+      '!extensions/amp-access/0.1/amp-login-done.js',
+      'builtins/**.js',
+      'third_party/caja/html-sanitizer.js',
+      'third_party/closure-library/sha384-generated.js',
+      'third_party/mustache/**/*.js',
+      'node_modules/document-register-element/build/' +
+          'document-register-element.max.js',
+      'node_modules/core-js/modules/**.js',
+      // Not sure what these files are, but they seem to duplicate code
+      // one level below and confuse the compiler.
+      '!node_modules/core-js/modules/library/**.js',
+      // Don't include tests.
+      '!**_test.js',
+      '!**/test-*.js',
+    ])
+    .pipe(closureCompiler({
+      // Temporary shipping with our own compiler that has a single patch
+      // applied
+      compilerPath: 'third_party/closure-compiler/compiler.jar',
+      fileName: intermediateFilename,
+      compilerFlags: {
+        // Custom compilation level. Trying to land this in the core
+        // compiler.
+        compilation_level: 'SIMPLE_PLUS_OPTIMIZATIONS',
+        // Transpile from ES6 to ES5.
+        language_in: 'ECMASCRIPT6',
+        language_out: 'ECMASCRIPT5',
+        js_module_root: ['node_modules/', 'build/fake-module/'],
+        common_js_entry_module: entryModuleFilename,
+        process_common_js_modules: true,
+        // This strips all files from the input set that aren't explicitly
+        // required.
+        only_closure_dependencies: true,
+        output_wrapper: wrapper,
+        create_source_map: intermediateFilename + '.map',
+        source_map_location_mapping: '|http://localhost:8000/'
+      }
+    }))
+    .on('error', function(err) {
+      if (/0 error\(s\)/.test(err.message)) {
+        // emit warning
+        console./*OK*/warn(err.message);
+        this.emit('end');
+      } else {
+        reject(err);
+      }
+    })
+    .on('end', function() {
+      console./*OK*/log('Compiled ', entryModuleFilename);
+      gulp.src(intermediateFilename + '.map')
+          .pipe(rename(outputFilename + '.map'))
+          .pipe(gulp.dest(outputDir))
+          .on('end', resolve);
+      return gulp.src(intermediateFilename)
```

Comment:
i'm perfectly fine with this as long as it works. was just worried if there was anything racy (because of async nature of tasks. guessing gulp knows its still working)

haha yes, gulp likes to make things that are supposed to be easy slightly hard sometimes..